dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/enrollment/AbstractEnrollmentService.java

```diff
@@ -530,7 +607,9 @@ private ImportSummary addEnrollment( Enrollment enrollment, ImportOptions import
 
         programInstanceService.addProgramInstance( programInstance, importOptions.getUser() );
 
-        importSummary = validateProgramInstance( program, programInstance, enrollment, importSummary );
+        sessionFactory.getCurrentSession().merge( daoTrackedEntityInstance );
```

Comment:
We add the program to the tei in this case. We use merge because it is not a session object. This way we prefetch objects not related to the session, avoid locks and still be able to update/persist them

Reply:
ok, I still think we should try to move any session-related work like this to the store, could we for example make a merge method in the store?